Feat/provisiond message bus

DylanVerstraete opened the pull request.

DylanVerstraete committed:
wip message bus client

DylanVerstraete committed:
fix redis connection, add reply method

DylanVerstraete committed:
improve

DylanVerstraete committed:
wip provision engine mbus

muhamadazmy commented on pkg/mbus/mbus.go:
```diff
@@ -0,0 +1,168 @@
+)
+
+type Message struct {
+	Version    int    `json:"version"`
```
the json names **MUST** match the ones from the actual msgbus protocol as defined in the repo

muhamadazmy commented on pkg/mbus/mbus.go:
```diff
@@ -0,0 +1,168 @@
+	Expiration int    `json:"expiration"`
+	Retry      int    `json:"retry"`
+	Data       string `json:"data"`
+	Twin_src   []int  `json:"twin_src"`
```
Already the CI is complaining that this name is not a conventional name, it should be TwinSrc instead (same for others)

muhamadazmy commented on pkg/mbus/mbus.go:
```diff
@@ -0,0 +1,168 @@
+	Err        string `json:"err"`
+}
+
+type Messagebus struct {
```
```suggestion
type MessageBus struct {
```

muhamadazmy commented on pkg/mbus/mbus.go:
```diff
@@ -0,0 +1,168 @@
+	pool    *redis.Pool
+}
+
+func New(port uint16, context context.Context) (*Messagebus, error) {
```
check Ci complaining about the arguments order, it's a convention to have `Context` as first argument

muhamadazmy commented on pkg/mbus/mbus.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+func New(port uint16, context context.Context) (*Messagebus, error) {
+	addr := fmt.Sprintf("tcp://[::]:%d", port)
```
Who said it's going to be tcp always? We probably gonna do it over unix socket, hence the New argument should actually accept a 'connection' address instead (uninx:/// or tcp://)

muhamadazmy commented on pkg/provision/mbus/mbus.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+const (
+	cmdDeploy = "msgbus.deploy"
```
Why the cmd has `msgbus` prefix ? i think this instead should be something like

`zos.deployment.deploy`
`zos.deployment.get`

`zos.network.config.blahblah` 
and so on.

muhamadazmy commented on pkg/provision/mbus/reservation.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+func (a *WorkloadsMessagebus) CreateOrUpdate(ctx context.Context, message mbus.Message, create bool) (interface{}, mw.Response) {
+	bytes, err := base64.RawStdEncoding.DecodeString(message.Data)
```
This depends on the user of the library to actually know the encoding used in the message. hence it requires good knowledge of the underlying protocol. What happens if in a new version of the bus we decided to move to a binary message format ?

A better way to do this is something like
```
bytes, err := message.GetPayload() //or GetData
```
and that one takes care of decoding the message correctly

muhamadazmy commented on pkg/provision/mbus/mbus.go:
```diff
@@ -0,0 +1,61 @@
+	return api, nil
+}
+
+func (w *WorkloadsMessagebus) Run() error {
```
I don't like many things about this method
- it (again) requires knowledge of the internal of the message bus, the "Message" structure, the streams, and so on.
- Also a call to StreamMessages is already a "blocking" call, it will work in an infinite loop to pull messages (which is fine) but what is not fine, that the 2nd call to StreamMessages will never happen because the first call will never return.
- Why do i have to define channels, and do select, and much much more things that can be optimized and abstracted by the library itself.
- For example, and HTTP library, doesn't require knowledge of how the headers or the http request parsing is happening, while u as a user can still "get" this information from the request object. 

In my opinion, the library should have the following Interface:
```go

bus := rmb.New(url) // url to redis endpoint

bus.Handle("zos.deployment.deploy", deployHandler)
bus.Handle("zos.deployment.get", getHandler)
bus.Handle("zos.deployment.delete", deleteHandler)

sub := bus.Subrouter("network")
sub.Handle("config.get") // this should be similar to `bus.Handle("network.config.get")` but is cleaner to define may be and allow later injection of middle ware for different parts

// finally
bus.Run()
```

a function then can be something like
```go
func getHandler(ctx context.Context, payload bytes[]) (interface{}, error) {
  // load your pay load to parse request params
  // do your things
 // if needed access extra messages fields from ctx, like
 message := rmb.GetMessage(ctx)
  // or others!
  return result, nil
}
```

DylanVerstraete committed:
resolve some issues

DylanVerstraete committed:
refactor message bus

DylanVerstraete committed:
small fix

DylanVerstraete committed:
fix CI

DylanVerstraete committed:
fix CI

DylanVerstraete committed:
improvement on message bus

DylanVerstraete committed:
refactor context messagebus

DylanVerstraete committed:
improvements

muhamadazmy commented on pkg/rmb/rmb.go:
```diff
@@ -0,0 +1,251 @@
+}
+
+// SendReply send a reply to the message bus with some data
+func (m *MessageBus) SendReply(message Message, data interface{}) error {
```
This doesn't have to be exposed

muhamadazmy commented on pkg/rmb/rmb.go:
```diff
@@ -0,0 +1,251 @@
+
+// PushMessage pushes a message to a topic
+// for testing purposes
+func (m *MessageBus) PushMessage(topic string, message Message) error {
```
Same here, this doesn't have to be public

DylanVerstraete replied:
for testing purposes right now, I have a testproject which uses this method to push some messages to the queue. I will remove this method before we merge this PR

muhamadazmy commented on pkg/rmb/rmb.go:
```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+func GetMessage(ctx context.Context) (*Message, error) {
```
This should panic instead of returning an error. because failure to get the message is a programming error, not a runtime error. A failure to get the message meas the MessageBus never injected the message on the context. hence panic is much better here. 

It will also make the handler code much easier

muhamadazmy commented on pkg/rmb/rmb.go:
```diff
@@ -0,0 +1,251 @@
+			}
+
+			var tKey twinKeyID
+			ctx = context.WithValue(ctx, tKey, message.TwinSrc)
```
this should be 
`requestCtx := context.WithValue(ctx, twinKeyID{}, message.TwinSrc)`
the idea is to create a new context per request. now this is instead overriding the context passed to the function because this is in a loop. so u actually override the ctx each time

Hence the correct way to do it is to create a new context for each request which leaves the ctx passed to the function as it is. this main ctx will be used for cancellation later

muhamadazmy commented on pkg/rmb/rmb.go:
```diff
@@ -0,0 +1,251 @@
+
+// Run runs listeners to the configured handlers
+// and will trigger the handlers in the case an event comes in
+func (m *MessageBus) Run() error {
```
```suggestion
func (m *MessageBus) Run(ctx context.Context) error {
```

muhamadazmy commented on pkg/rmb/rmb.go:
```diff
@@ -0,0 +1,251 @@
+
+	jobs := make(chan Message, numWorkers)
+	for i := 1; i <= numWorkers; i++ {
+		go m.worker(context.Background(), jobs)
```
```suggestion
		go m.worker(ctx, jobs)
```

muhamadazmy commented on pkg/rmb/rmb.go:
```diff
@@ -0,0 +1,251 @@
+			continue
+		}
+
+		jobs <- message
```
while this should work but there is possible loss of messages that can happen if a message is popped from the queue but then this code here was stopped before the message get chance to be processed. I am not sure if this is something that we should handle ourselves or by the rmb itself! @maxux ?

DylanVerstraete committed:
resolve comments

DylanVerstraete committed:
fix CI

DylanVerstraete committed:
fix CI

DylanVerstraete committed:
port network and statistics api to rmb

DylanVerstraete committed:
add startup

muhamadazmy committed:
wip: support subrouters

muhamadazmy committed:
adding some tests

muhamadazmy committed:
no need to check handler in main

muhamadazmy committed:
fixes + update tests

muhamadazmy committed:
allow chaining by returning same subrouter

DylanVerstraete committed:
cleanup routes

DylanVerstraete committed:
fix and refactors

DylanVerstraete committed:
some fixes for deployments rmb

DylanVerstraete committed:
connect to redis socket

DylanVerstraete committed:
fix CI

muhamadazmy committed:
Separate the API

also improve the message bus retry mechanism

muhamadazmy committed:
ci fixes

muhamadazmy committed:
more ci fixes

muhamadazmy committed:
fix reply queue

muhamadazmy committed:
hide key, and type conversion from user

User of the library does not need to know about keys, or types
and not having the twinid in ctx is a programming
error, not runtime, hence a panic is raised, not
returning a bad request

muhamadazmy committed:
WIP: bus client

muhamadazmy committed:
use twin id from message

muhamadazmy committed:
Working test

muhamadazmy committed:
reduce timeout time to 3

muhamadazmy committed:
ci

muhamadazmy committed:
protect the public_config_set method for now

muhamadazmy committed:
add more methods to client
